v0/destinations/mailchimp/utils.js

```diff
@@ -0,0 +1,249 @@
+  } else {
+    mergeFieldOption = enableMergeFields;
+  }
+  // ref: https://mailchimp.com/developer/marketing/api/lists/batch-subscribe-or-unsubscribe/
+  const BASE_URL = `https://${datacenterId}.api.mailchimp.com/3.0/lists/${audienceId}`;
+
+  const BATCH_ENDPOINT = `${BASE_URL}?skip_merge_validation=${mergeFieldOption}&skip_duplicate_check=false`;
+
+  return BATCH_ENDPOINT;
+};
+
+/**
+ * Converts object keys to upper case along with removing spaces
+ * It also slices Keys more than 10 characters
+ * @param {*} tag <-- object key
+ * @param {*} returns
+ */
+
+const filterTagValue = tag => {
+  const maxLength = 10;
+  let newTag = tag.replace(/[^\w\s]/gi, "");
+  if (newTag.length > maxLength) {
+    newTag = newTag.slice(0, 10);
+  }
+  return newTag.toUpperCase();
+};
+
+/**
+ * Returns true if the email id already exists in mailchimp database
+ * @param {*} apiKey <-- from webapp config
+ * @param {*} datacenterId <-- from webapp config
+ * @param {*} audienceId <-- from webapp config
+ * @param {*} email <-- from webapp config
+ * @param {*} returns
+ */
+const checkIfMailExists = async (apiKey, datacenterId, audienceId, email) => {
+  // ref: https://mailchimp.com/developer/marketing/api/list-members/get-member-info/
+  let status = false;
+  if (!email) {
+    return false;
+  }
+  const hash = md5(email);
+  const url = `${getMailChimpEndpoint(
+    datacenterId,
+    audienceId
+  )}/members/${hash}`;
+  const basicAuth = Buffer.from(`apiKey:${apiKey}`).toString("base64");
+
+  try {
+    response = await axios.get(url, {
+      headers: {
+          Authorization: `Basic ${basicAuth}`,
+      },
+  });
+    status = true;
+  } catch (error) {
+    logger.error("axios error");
+  }
+  return status;
+};
+
+/**
+ * Checks if double optin setup is switched on for the particular audience id in mailchimp
+ * @param {*} apiKey <-- from webapp config
+ * @param {*} datacenterId <-- from webapp config
+ * @param {*} audienceId <-- from webapp config
+ * @param {*} returns
+ */
+const checkIfDoubleOptIn = async (apiKey, datacenterId, audienceId) => {
+  let doubleOptin;
+  const url = `${getMailChimpEndpoint(datacenterId, audienceId)}`;
+  const basicAuth = Buffer.from(`apiKey:${apiKey}`).toString("base64");
+  try {
+    response = await axios.get(url, {
+        headers: {
+            Authorization: `Basic ${basicAuth}`,
+        },
+    });
+} catch (error) {
+  throw new CustomError(
+          "[Mailchimp]:: User does not have access to the requested operation",
+             error.status || 400
+           );
+}
+return !!doubleOptin;
+};
+
+/**
+ * Deduces final endpoint and method for existing emails in mailchimp
+ * @param {*} datacenterId <-- from webapp config
+ * @param {*} audienceId <-- from webapp config
+ * @param {*} email <-- email value from input payload
+ * @param {*} response <-- default rudder response structure
+ * @param {*} returns
+ */
+const stitchEndpointAndMethodForExistingEmails = (
```

Comment:
This is add and update endpoint we can only use this rather that using 2 endpoints for 2 separate functions